File: backend/app/schemas/analytics_qa.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.custom_reports import ReportDatasetFieldRead, ReportQueryConfig


class AnalyticsQaQuestionRequest(BaseModel):
    question: str = Field(min_length=1, max_length=500)


class AnalyticsQaChartSuggestionRead(BaseModel):
    type: str
    x_field: str | None = None
    y_field: str | None = None


class AnalyticsQaAnswerRead(BaseModel):
    question: str
    answer: str
    source: Literal["ai", "rules"]
    query_config: ReportQueryConfig
    columns: list[ReportDatasetFieldRead]
    rows: list[dict[str, Any]]
    row_count: int
    chart: AnalyticsQaChartSuggestionRead

File: frontend/src/api/analyticsQa.ts
import type { AxiosInstance } from 'axios';

import { apiClient } from '@/api/client';
import type { ReportDatasetFieldRead, ReportQueryConfig } from '@/types/customReports';

interface ApiResponse<T> {
  code: number;
  message: string;
  data: T;
}

export interface AnalyticsQaChartSuggestion {
  type: string;
  x_field: string | null;
  y_field: string | null;
}

export interface AnalyticsQaAnswerRead {
  answer: string;
  chart: AnalyticsQaChartSuggestion;
  columns: ReportDatasetFieldRead[];
  query_config: ReportQueryConfig;
  question: string;
  row_count: number;
  rows: Record<string, unknown>[];
  source: 'ai' | 'rules';
}

export async function askAnalyticsQuestion(
  question: string,
  client: AxiosInstance = apiClient,
): Promise<AnalyticsQaAnswerRead> {
  const response = await client.post<ApiResponse<AnalyticsQaAnswerRead>>('/analytics-qa/ask', {
    question,
  });
  return response.data.data;
}
